### PR TITLE
Parallel Gather Satellite Graphs

### PR DESCRIPTION
**Pull request by maierlars**

### CODE REVIEW

File: arangod/Aql/ClusterNodes.cpp [hunk 1]
```diff
@@ -589,21 +589,35 @@ struct ParallelizableFinder final : public WalkerWorker<ExecutionNode> {
   }
 
   bool before(ExecutionNode* node) override final {
-    if (node->getType() == ExecutionNode::SCATTER || node->getType() == ExecutionNode::GATHER ||
-        node->getType() == ExecutionNode::DISTRIBUTE ||
-        node->getType() == ExecutionNode::TRAVERSAL ||
-        node->getType() == ExecutionNode::SHORTEST_PATH ||
-        node->getType() == ExecutionNode::K_SHORTEST_PATHS) {
+    LOG_DEVEL << "ParallelizableFinder::before " << node->id();
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]:
```suggestion
```

File: arangod/Aql/ClusterNodes.cpp [hunk 2]
```diff
@@ -589,21 +589,35 @@ struct ParallelizableFinder final : public WalkerWorker<ExecutionNode> {
...
+    auto nodeType = node->getType();
+
+    if (nodeType == ExecutionNode::SCATTER || nodeType == ExecutionNode::GATHER ||
+        nodeType == ExecutionNode::DISTRIBUTE) {
+      LOG_DEVEL << "not parallelizable (1)";
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]:
```suggestion
```

File: arangod/Aql/ClusterNodes.cpp [hunk 3]
```diff
@@ -589,21 +589,35 @@ struct ParallelizableFinder final : public WalkerWorker<ExecutionNode> {
...
       _isParallelizable = false;
       return true;  // true to abort the whole walking process
     }
+
+    if (nodeType == ExecutionNode::TRAVERSAL || nodeType == ExecutionNode::SHORTEST_PATH ||
+        nodeType == ExecutionNode::K_SHORTEST_PATHS) {
+      LOG_DEVEL << "traversal node";
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]:
```suggestion
```

File: arangod/Aql/ClusterNodes.cpp [hunk 4]
```diff
@@ -589,21 +589,35 @@ struct ParallelizableFinder final : public WalkerWorker<ExecutionNode> {
...
+      GraphNode* gn = ExecutionNode::castTo<GraphNode*>(node);
+      if (!gn->isUsedAsSatellite()) {
+        LOG_DEVEL << "graph node not used as satellite";
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]:
```suggestion
```

File: arangod/Aql/ClusterNodes.cpp [hunk 5]
```diff
@@ -589,21 +589,35 @@ struct ParallelizableFinder final : public WalkerWorker<ExecutionNode> {
...
+      LOG_DEVEL << "not parallelizable (1)";
       _isParallelizable = false;
       return true;  // true to abort the whole walking process
     }
+
+    if (nodeType == ExecutionNode::TRAVERSAL || nodeType == ExecutionNode::SHORTEST_PATH ||
+        nodeType == ExecutionNode::K_SHORTEST_PATHS) {
+      LOG_DEVEL << "traversal node";
+      GraphNode* gn = ExecutionNode::castTo<GraphNode*>(node);
+      if (!gn->isUsedAsSatellite()) {
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]: As discussed, it suffices to test whether the `GraphNode` is also a `SatelliteGraphNode`. e.g
```suggestion
      auto* gn = ExecutionNode::castTo<GraphNode*>(node);
      if (!gn->isSatelliteNode()) {
```

File: arangod/Aql/ClusterNodes.cpp [hunk 6]
```diff
@@ -589,21 +589,35 @@ struct ParallelizableFinder final : public WalkerWorker<ExecutionNode> {
...
+    }
+
     // write operations of type REMOVE, REPLACE and UPDATE
     // can be parallelized, provided the rest of the plan
     // does not prohibit this
     if (node->isModificationNode() &&
         (!_parallelizeWrites || (node->getType() != ExecutionNode::REMOVE &&
                                  node->getType() != ExecutionNode::REPLACE &&
                                  node->getType() != ExecutionNode::UPDATE))) {
+      LOG_DEVEL << "not parallelizable ()";
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]:
```suggestion
```

File: arangod/Aql/ClusterNodes.cpp [hunk 7]
```diff
@@ -620,8 +634,11 @@ bool GatherNode::isParallelizable() const {
     return false;
   }
 
+  LOG_DEVEL << "checking is all dependencies of gather are parallelizable";
+
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]:
```suggestion
```

File: arangod/Aql/ClusterNodes.cpp [hunk 8]
```diff
@@ -620,8 +634,11 @@ bool GatherNode::isParallelizable() const {
...
   ParallelizableFinder finder(*_vocbase);
   for (ExecutionNode* e : _dependencies) {
+    LOG_DEVEL << "walking node " << e->id();
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]:
```suggestion
```

File: arangod/Aql/OptimizerRules.cpp [hunk 9]
```diff
@@ -7362,14 +7362,23 @@ void arangodb::aql::parallelizeGatherRule(Optimizer* opt,
   ::arangodb::containers::SmallVector<ExecutionNode*> nodes{a};
   plan->findNodesOfType(nodes, EN::GATHER, true);
 
-  if (nodes.size() == 1 && !plan->contains(EN::TRAVERSAL) &&
-      !plan->contains(EN::SHORTEST_PATH) && !plan->contains(EN::K_SHORTEST_PATHS) &&
-      !plan->contains(EN::DISTRIBUTE) && !plan->contains(EN::SCATTER)) {
+  if (nodes.size() == 1 && !plan->contains(EN::DISTRIBUTE) && !plan->contains(EN::SCATTER)) {
     GatherNode* gn = ExecutionNode::castTo<GatherNode*>(nodes[0]);
+    LOG_DEVEL << "Plan could be parallelized";
```
Comment (goedderz) [inline arangod/Aql/OptimizerRules.cpp]:
```suggestion
```

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9142/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9143/

**Pull request merged**